File: pickleball/pickleball/src/main/java/com/pickleball_backend/pickleball/service/GroupService.java
package com.pickleball_backend.pickleball.service;

import com.pickleball_backend.pickleball.entity.Group;
import com.pickleball_backend.pickleball.entity.GroupMember;
import com.pickleball_backend.pickleball.entity.GroupMessage;
import com.pickleball_backend.pickleball.entity.User;

import java.util.List;

public interface GroupService {
    
    // Group management
    Group createGroup(String name, String description, User creator, List<String> memberUsernames);
    Group getGroupById(Long groupId);
    List<Group> getUserGroups(Integer userId);
    Group updateGroup(Long groupId, String name, String description);
    void deleteGroup(Long groupId, Integer userId);
    
    // Member management
    void addMemberToGroup(Long groupId, String username, Integer userId);
    void removeMemberFromGroup(Long groupId, String username, Integer userId);
    void makeMemberAdmin(Long groupId, String username, Integer userId);
    void removeMemberAdmin(Long groupId, String username, Integer userId);
    List<GroupMember> getGroupMembers(Long groupId);
    boolean isUserMemberOfGroup(Long groupId, Integer userId);
    boolean isUserCreatorOfGroup(Long groupId, Integer userId);
    boolean isUserAdminOfGroup(Long groupId, Integer userId);
    
    // Message management
    GroupMessage sendMessage(Long groupId, String content, String imageUrl, Integer senderId);
    List<GroupMessage> getGroupMessages(Long groupId, int page, int size);
    void markMessagesAsRead(Long groupId, Integer userId, Long lastReadMessageId);
    void deleteMessage(Long messageId, Integer userId);
    void editMessage(Long messageId, String newContent, Integer userId);
    
    // Search
    List<Group> searchGroups(String searchTerm);
    List<Group> searchUserGroups(String searchTerm, Integer userId);
}

File: pickleball-frontend/src/service/GroupService.js
import api from './api';

const GroupService = {
  // Create a new group
  createGroup: async (groupName, description, memberUsernames) => {
    try {
      console.log('GroupService: Sending request to create group:', {
        groupName,
        description,
        memberUsernames
      });
      
      const response = await api.post('/groups/create', {
        groupName,
        description,
        memberUsernames
      });
      
      console.log('GroupService: Response received:', response.data);
      return response.data;
    } catch (error) {
      console.error('GroupService: Error creating group:', error);
      console.error('GroupService: Error response:', error.response?.data);
      throw error;
    }
  },

  // Get user's groups
  getUserGroups: async () => {
    try {
      const response = await api.get('/groups/my-groups');
      return response.data;
    } catch (error) {
      console.error('Error fetching user groups:', error);
      throw error;
    }
  },

  // Get group by ID
  getGroupById: async (groupId) => {
    try {
      const response = await api.get(`/groups/${groupId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching group:', error);
      throw error;
    }
  },

  // Get group members
  getGroupMembers: async (groupId) => {
    try {
      const response = await api.get(`/groups/${groupId}/members`);
      return response.data;
    } catch (error) {
      console.error('Error fetching group members:', error);
      throw error;
    }
  },

  // Add member to group
  addMemberToGroup: async (groupId, username) => {
    try {
      const response = await api.post(`/groups/${groupId}/members`, {
        username
      });
      return response.data;
    } catch (error) {
      console.error('Error adding member to group:', error);
      throw error;
    }
  },

  // Remove member from group
  removeMemberFromGroup: async (groupId, username) => {
    try {
      const response = await api.delete(`/groups/${groupId}/members/${username}`);
      return response.data;
    } catch (error) {
      console.error('Error removing member from group:', error);
      throw error;
    }
  },

  // Get group messages
  getGroupMessages: async (groupId, page = 0, size = 50) => {
    try {
      const response = await api.get(`/groups/${groupId}/messages`, {
        params: { page, size }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching group messages:', error);
      throw error;
    }
  },

  // Send message to group
  sendMessage: async (groupId, content, imageUrl = null) => {
    try {
      const response = await api.post(`/groups/${groupId}/messages`, {
        content,
        imageUrl
      });
      return response.data;
    } catch (error) {
      console.error('Error sending message:', error);
      throw error;
    }
  },

  // Mark messages as read
  markMessagesAsRead: async (groupId, lastReadMessageId) => {
    try {
      const response = await api.post(`/groups/${groupId}/messages/read`, {
        lastReadMessageId
      });
      return response.data;
    } catch (error) {
      console.error('Error marking messages as read:', error);
      throw error;
    }
  },

  // Delete group
  deleteGroup: async (groupId) => {
    try {
      const response = await api.delete(`/groups/${groupId}`);
      return response.data;
    } catch (error) {
      console.error('Error deleting group:', error);
      throw error;
    }
  },

  // Search groups
  searchGroups: async (query) => {
    try {
      const response = await api.get('/groups/search', {
        params: { query }
      });
      return response.data;
    } catch (error) {
      console.error('Error searching groups:', error);
      throw error;
    }
  },

  // Send group invitations (legacy endpoint)
  sendGroupInvitations: async (groupName, memberUsernames) => {
    try {
      const response = await api.post('/groups/send-invitations', {
        groupName,
        memberUsernames
      });
      return response.data;
    } catch (error) {
      console.error('Error sending group invitations:', error);
      throw error;
    }
  }
};

export default GroupService;
